New Dashboard Primitive - Headings

JesseSDevaney opened the pull request:
**Problem** 
Managing vertical space is exhausting on dashboards. The setup for sections and headings requires too many steps for such a simple requirement. 

If you want to create sections, you have to manually: 
1. create a text dashcard
2. resize it to full width and single cell height
3. leave no text in it
4. toggle the background to off. 

If you want to create headings, you have to manually:
1. create a text dashcard
2. resize it to full width and single cell height
3. remember to throw in a markdown `##` for h2 headings
4. toggle the background off.

---
**Context**
- Product Doc: https://www.notion.so/metabase/Create-a-new-heading-dashboard-primitive-b2e00171221849df9346dd47ed953803?pvs=4
- Closes: https://github.com/metabase/metabase/issues/30015
---
**Desired Behavior** 
- [ ] Ability to create a `Heading` primitive on dashboards
  - backed by an h2 element
  - placeholder: "Heading"
  - default to 100% wide
  - lets a user type in regular text to form a heading (no markdown required)
  - no background, no viz settings
  - auto-edit on click, auto-preview when not focused (edit mode)
  - auto-focus input on creation
- [ ] Create heading + text dropdown selector
  - add tooltip on hover: "Add a heading or text"
- [ ] Update text box
  - auto-focus textarea on creation
  - auto-edit when clicked, auto-preview when not focused (edit mode)
    - removed preview/edit mode toggle
- [ ] Style text and headings
  - blue border on hover/focus
  - blue border and blue background on drag
  - blue border when no content in text or headings


---
**Solution Implementation**

_Button Tooltip + Menu_
![button-tooltip-and-menu](https://github.com/metabase/metabase/assets/22608765/ba90ca09-1899-4ce3-b9a2-ceb47fc179b9)



_New Heading DashCard_
![new-heading-primitive](https://github.com/metabase/metabase/assets/22608765/51360be6-14c9-4327-be41-16ab881144f5)



_Text DashCard Updates_
![text-dashcard-updates](https://github.com/metabase/metabase/assets/22608765/83c412ee-5f6f-45b4-a495-155b8f8c787e)

_Example Usage_
![all-together](https://github.com/metabase/metabase/assets/22608765/a6cb5f52-7fc0-4950-bded-6fdd565120d0)

---
**Acceptance Criteria**
- [ ] Implementation behaves according to the desired behavior
- [ ] New tests pass, and other tests continue to pass

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30804)
<!-- Reviewable:end -->

JesseSDevaney committed:
initial prototype commit

JesseSDevaney committed:
add default value for text cards

JesseSDevaney committed:
auto focus text/heading cards on creation

JesseSDevaney committed:
remove default value - use placeholders instead

JesseSDevaney committed:
fix inability to edit when there is no heading value

JesseSDevaney committed:
remove unnecessary comment

JesseSDevaney committed:
re-word heading & text tooltip

JesseSDevaney committed:
refactor and fix draggable area bug

JesseSDevaney committed:
fix border artifacts when combined with dragging styling

JesseSDevaney committed:
Merge branch 'master' into new-dashboard-primitive-headings-prototype

JesseSDevaney committed:
fix stroke width for hovering and focused - make them the same

JesseSDevaney committed:
display placeholder value in editing mode if content is empty for text and heading dash-cards

JesseSDevaney committed:
remove duplicate placeholders

JesseSDevaney committed:
allow pointer-events anywhere on the div (not just content)

JesseSDevaney committed:
refactor styling to apply no content borders

JesseSDevaney committed:
Fix heading overflow

JesseSDevaney committed:
refactor text dash-card button

- remove unnecessary redux store usage

JesseSDevaney committed:
refactor border styling to one css format

JesseSDevaney committed:
rename class names to match syntax conventions

JesseSDevaney committed:
refactor auto-preview styling for dash-card root

JesseSDevaney committed:
refactor style name

JesseSDevaney committed:
remove comment

JesseSDevaney committed:
undo unnecessary re-formatting

JesseSDevaney committed:
refactor popover styling

JesseSDevaney committed:
remove comment that was previously addressed

JesseSDevaney committed:
refactor to typescript

JesseSDevaney committed:
remove comment

JesseSDevaney committed:
remove un-used code

JesseSDevaney committed:
fix TypeScript errors in text unit spec

JesseSDevaney committed:
update button key and event description

JesseSDevaney committed:
add e2e tests for text and heading dash-cards

JesseSDevaney committed:
add e2e tests

JesseSDevaney committed:
add small heading unit tests

JesseSDevaney committed:
fix typescript error

JesseSDevaney committed:
Merge branch 'master' into new-dashboard-primitive-headings-prototype

JesseSDevaney committed:
add aria-label removed by merging with master

JesseSDevaney committed:
fix broken test

JesseSDevaney committed:
fix broken tests

JesseSDevaney committed:
fix typescript errors

JesseSDevaney committed:
Merge branch 'master' into new-dashboard-primitive-headings-prototype

JesseSDevaney committed:
align edit text with preview text

JesseSDevaney committed:
Merge branch 'master' into new-dashboard-primitive-headings-prototype

alxnddr commented on frontend/src/metabase/css/dashboard.css:
```diff
@@ -162,6 +162,27 @@
   box-shadow: 3px 3px 8px var(--color-shadow);
 }
 
+.TextCard .has-no-content {
```
Is it possible to move these styles to Text/Heading components? Seems like `.TextCard` class is assigned in the `DashboardGrid` in `renderGridItem` which sort of broadens its responsibility to not only being just the layout of abstract cards but also knowing what cards are and which classes we should assign

JesseSDevaney replied:
The reason these are not in Text/Heading components themselves is that `react-grid-layout` applies styles to its components via CSS classes. Certain styles are only customizable by targeting the CSS class applied to the them by `react-grid-layout`.

So, in order to change the color of the resizing handle
![image](https://github.com/metabase/metabase/assets/22608765/f5046825-c226-4b94-8601-e31e55c38a0c)
and styling applied during "resizing"
![resizing-styling](https://github.com/metabase/metabase/assets/22608765/1d30a66e-f36a-4497-96f2-2375a23311b4)
we need to be able to target `.react-resizable-handle:after` and `.resizing`. The `.TextCard` class is chained before them to further specify that we want to apply these styles only to text/heading cards - not other types of dash-cards. 

Also I used
```css
.TextCard:hover .text-edit-container,
.TextCard:focus-within .text-edit-container {
  border-radius: 8px;
  border: 1px solid var(--color-brand);
}
```
because `react-grid-layout` was consuming/stopping the propagation of hover events over its draggable surface area such that only hovering over the inner-portion of a card applied correct border styling, but we want the whole card to display a border when hovered.

Of course, this below could be moved to the text/heading components themselves b/c it is not affected by `react-grid-layout`, but then we would have this kind of styling in multiple places which seemed like a bad/confusing idea to me.
```css
.TextCard .has-no-content {
  border-radius: 8px;
  border: 1px solid var(--color-brand);
}
```

alxnddr replied:
We could eliminate `text-edit-container` global class and move the logic to the `Heading` component. If it needs to be reusable between Text and Heading cards, we could extract a `TextCardContainer`. But yes, to style the drag handle we need some global css

ts```
import { css } from "@emotion/react";
import styled from "@emotion/styled";
import { color } from "metabase/lib/colors";

interface InputContainerProps {
  isPreviewing: boolean;
  isEmpty: boolean;
}

export const InputContainer = styled.div<InputContainerProps>`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  justify-content: center;
  overflow: hidden;
  padding: 0 0.65em;
  padding-left: 0;
  pointer-events: auto;
  border-radius: 8px;

  .DashCard:hover &,
  .DashCard:focus-within & {
    border: 1px solid ${color("brand")};
  }

  .DashCard.react-draggable-dragging &,
  .DashCard.react-resizable-resizing & {
    border: unset;
  }

  ${({ isPreviewing }) =>
    !isPreviewing &&
    css`
      padding: 0 0.75em;
    `}

  ${({ isEmpty }) =>
    isEmpty &&
    css`
      border: 1px solid ${color("brand")};
    `}
`;
```

alxnddr commented on frontend/src/metabase/dashboard/components/DashCard/DashCard.styled.tsx:
```diff
@@ -6,6 +6,7 @@ export interface DashCardRootProps {
   isNightMode: boolean;
   isUsuallySlow: boolean;
   hasHiddenBackground: boolean;
+  shouldAutoPreviewInEditing: boolean;
```
This is a generic DashCard which can be a line chart that does not have preview/editing states so it is not a generic attribute. I'd suggest moving all text cards' specific logic inside their components

JesseSDevaney replied:
This is placed inside the generic DashCard b/c `box-shadow` and `background` are DashCard applied styles. If we want Text/Heading components to be able to auto-preview, we need to be able to change the DashCard styling that gets applied to them while they are editing. Otherwise, if we were to auto-preview 
1. a `<Heading />` component or 
2. a `<Text />` component with background disabled

it would still show a white background and box-shadow around the card since those are generically applied to DashCards.

**With DashCard specific styling**
![image](https://github.com/metabase/metabase/assets/22608765/93038cde-b68f-46a3-9e11-648d9ab34928)

**Without DashCard styling**
![image](https://github.com/metabase/metabase/assets/22608765/7a0f9399-ee6e-4e1f-a520-01eeeb468fb8)

As you can see heading components and text components w/ background disabled should never show a background or have a box-shadow, but they do if the styling is not applied to the DashCard component.

If you have a idea that solves this issue, please let me know :heart:

alxnddr commented on frontend/src/metabase/dashboard/components/TextOptions/TextChoicesPopover.tsx:
```diff
@@ -0,0 +1,52 @@
+import React from "react";
```
We have a menu like this in other places like the "New" button, wdyt about reusing it?
<img width="232" alt="Screen Shot 2023-05-25 at 9 12 07 PM" src="https://github.com/metabase/metabase/assets/14301985/203963e1-140e-4c1e-9cc0-153c214910ef">

JesseSDevaney replied:
Thanks for pointing this out!
I will test this out and get designs feedback for it.

alxnddr commented on frontend/src/metabase/visualizations/visualizations/Heading/Heading.tsx:
```diff
@@ -0,0 +1,76 @@
+interface HeadingProps {
+  isEditing: boolean;
+  onUpdateVisualizationSettings: ({ text }: { text: string }) => void;
+  dashcard: { justAdded?: boolean };
```
You can use `BaseDashboardOrderedCard` from `frontend/src/metabase-types/api/dashboard.ts`. And for settings on the line below we have `VisualizationSettings` from `frontend/src/metabase-types/api/card.ts`

alxnddr commented on frontend/src/metabase/visualizations/visualizations/Heading/Heading.unit.spec.tsx:
```diff
@@ -0,0 +1,42 @@
+  onUpdateVisualizationSettings: ({ text }: { text: string }) => {
+    return;
+  },
+  dashcard: {},
```
To simplify writing specs we have mock object factories like `createMockDashboardCardWithVirtualCard` in `frontend/src/metabase-types/api/mocks/dashboard.ts` — you will need it once you update the props types

alxnddr commented on frontend/src/metabase/visualizations/visualizations/Text/Text.styled.jsx:
```diff
@@ -0,0 +1,52 @@
+  box-shadow: none;
+  font-size: 1.143em;
+  height: inherit;
+  line-height: 1.602em;
```
Here and in other places you added very precise and unusual values we don't have anywhere else. Usually, we use `rem` with the step of `0.25` (rarely `0.125`)

JesseSDevaney replied:
These specific values were just ported over from converting part of this file to styled-components from CSS modules. I did not choose these values in particular as they already existed, but I will check other places I created new styles and update appropriately.

Thank you for the heads up!

alxnddr reviewed (commented):
Great progress 👏 

We also need to update `getSupportedCardsForSubscriptions` where we exclude "virtual" cards.

Is it expected that I can change the width of header cards?
<img width="927" alt="Screen Shot 2023-05-25 at 9 23 56 PM" src="https://github.com/metabase/metabase/assets/14301985/c88f9197-f2d5-4d03-939b-5e5a7a71ffdb">

Also, we need to tweak styles for the mobile resolutions:
https://github.com/metabase/metabase/assets/14301985/164c63f6-0959-4dfe-b83f-9a83475972f1

JesseSDevaney commented on frontend/src/metabase/visualizations/visualizations/Text/Text.jsx:
```diff
@@ -25,188 +31,113 @@ const REHYPE_PLUGINS = [
+          "has-no-content": hasNoContent,
         })}
+        isPreviewing={isPreviewing}
+        onMouseEnter={() => setIsHovering(true)}
```
I checked with the designers and they prefer click editing over hover editing.
Thank you for the suggestion! Also, passing along a thanks from Conor and Nathlia! :heart:

JesseSDevaney replied:
**Hover Editing**
![hover-edit](https://github.com/metabase/metabase/assets/22608765/9f206abf-f979-4764-afdc-a7eb8ab93c6d)
![hover-edit-all](https://github.com/metabase/metabase/assets/22608765/24f23baa-bf49-4a9c-80b7-5525c665961e)



**Click Editing**
![click-edit](https://github.com/metabase/metabase/assets/22608765/ab31e92c-0501-4ad9-9e65-7557fad1a5a2)
![click-edit-all](https://github.com/metabase/metabase/assets/22608765/c6192ff2-02d1-48cf-8648-99e32e12f1b4)

JesseSDevaney commented:
> Also, we need to tweak styles for the mobile resolutions: https://github.com/metabase/metabase/assets/14301985/164c63f6-0959-4dfe-b83f-9a83475972f1

@alxnddr Could you repost this media? It is not letting me preview or watch the image/video.

JesseSDevaney committed:
change from hover editing to click editing

JesseSDevaney committed:
update tests for click editing

- remove hover tests
- change focus tests to click tests
- remove top-level findByText() calls

JesseSDevaney committed:
refactor to use-focus hook

JesseSDevaney committed:
add "heading" to getSupportedCardsForSubscriptions

JesseSDevaney committed:
rename showVisualizationOptions to showDashcardVisualizationSettings

JesseSDevaney committed:
add aria label to buttons and query in tests by label

JesseSDevaney committed:
refactor E2E tests to make longer flows

JesseSDevaney committed:
add existing metabase types to Heading

JesseSDevaney committed:
[skip ci] Merge branch 'master' into new-dashboard-primitive-headings-prototype

JesseSDevaney committed:
refactor heading and text menu to use EntityMenu

JesseSDevaney committed:
fix aria-label

JesseSDevaney committed:
[skip ci] fix css syntax error from missing semi-colon

JesseSDevaney committed:
adjust styles and use rem value standards

JesseSDevaney committed:
[skip ci] prevent drag on click to edit surface area

iethree commented on frontend/src/metabase/hooks/use-focus.ts:
correct me if i'm wrong, but this hook seems to do the same thing as our existing `use-toggle` hook, right? If so, we should use that one.

https://github.com/metabase/metabase/blob/ffc2db4932d6affc794eae85c6fddf49acd78252/frontend/src/metabase/hooks/use-toggle.ts#L8-L18

there's also the `useToggle` from react-use library, which we already have
https://github.com/streamich/react-use/blob/master/docs/useToggle.md

iethree commented on frontend/src/metabase/visualizations/visualizations/Heading/Heading.tsx:
```diff
@@ -0,0 +1,84 @@
+            isEditing={isEditing}
+            onMouseDown={preventDragging}
+          >
+            {content || placeholder}
```
are we sure we'll never get a numeric zero here? If not, maybe try using `isEmpty` ?

https://github.com/metabase/metabase/blob/ffc2db4932d6affc794eae85c6fddf49acd78252/frontend/src/metabase/lib/validate.js#L6-L7


```suggestion
            {isEmpty(content) ? content : placeholder}
```

JesseSDevaney committed:
tweak styles for mobile resolutions

iethree commented on frontend/src/metabase/visualizations/visualizations/Text/Text.jsx:
```diff
@@ -25,188 +34,115 @@ const REHYPE_PLUGINS = [
+  const isPreviewing = !isFocused;
+
+  const handleTextChange = text =>
+    onUpdateVisualizationSettings({ text: text });
```
nit

```suggestion
    onUpdateVisualizationSettings({ text });
```

iethree commented on frontend/src/metabase/visualizations/visualizations/Text/Text.jsx:
```diff
@@ -25,188 +34,115 @@ const REHYPE_PLUGINS = [
+    onUpdateVisualizationSettings({ text: text });
+  const preventDragging = e => e.stopPropagation();
+
+  const isSingleRow = gridSize && gridSize.height === 1;
```
nit

```suggestion
  const isSingleRow = gridSize?.height === 1;
```

JesseSDevaney commented:
@iethree @alxnddr I addressed all feedback besides the global css and DashCard styling - since there is not a clear improvement path yet, I left it marked unresolved above. It is ready for re-review whenever you get the chance. 

@alxnddr 
1. The ability to the change the width of a Heading component is expected - we just want it to default to 100% wide.
2. I decreased the height of heading cards in mobile resolutions, but it appears there is another issue where all charts get moved to the top leaving headings and text in areas with no context.
**Desktop Resolution**
![image](https://github.com/metabase/metabase/assets/22608765/35242823-8674-4255-bd11-f016c7e412ef)
**Mobile Resolution**
![image](https://github.com/metabase/metabase/assets/22608765/6f9bd00c-1a10-439f-a54d-f4fed2e1fcf4)



Thanks for the help and suggestions - it has significantly improved this PR. Let me know if there are any other concerns we should address :heart:

JesseSDevaney committed:
fix react import lint error

iethree reviewed (changes requested):
🏆  This is really great work. The feature is great, you understand the requirements well, and the code is excellent.

I've added a lot of comments, 99% of which are nits, or things that should be changed to fit our existing practices and styles. Hopefully they're informative and helpful in learning about tools in our codebase as well as some of our testing practices.

One more general comment, this PR is pretty big. Don't bother breaking it down now, but it could have been easier to write + review if you had maybe broken it down into separate PRs as follows:

1. Convert Text visualization to functional component.
2. Convert Text visualization to typescript
3. Add auto-preview behavior to text visualization.
4. Add Text options button (with both buttons just adding text cards)
5. Add Header visualization

(I also find it's much easier to get prompt reviews when PRs are smaller 😉  )

_PS. I noticed you added a few commit while working on this. If the PR isn't ready for review, move it back to draft status._

JesseSDevaney committed:
fix E2E tests

JesseSDevaney committed:
refactor e2e helper to shorthand

JesseSDevaney committed:
refactor to use e2e helper

JesseSDevaney committed:
refactor to use saveDashboard() e2e helper

JesseSDevaney commented on e2e/test/scenarios/dashboard/text-dashcards.cy.spec.js:
```diff
@@ -0,0 +1,202 @@
+      // Reload page
+      cy.reload();
+
+      getDashboardCard(1).within(() => {
+        cy.get("div").contains("Text text text");
+      });
```
findByText was not working here

JesseSDevaney replied:
findByText was not working here because react markdown styles the text with `<em>` and `<strong>` tags

JesseSDevaney commented on e2e/test/scenarios/dashboard/text-dashcards.cy.spec.js:
```diff
@@ -0,0 +1,202 @@
+        cy.get("textarea").should("have.focus");
+
+        // textarea editor should have placeholder "Heading"
+        cy.get("textarea").should(
+          "have.attr",
+          "placeholder",
+          "You can use Markdown here, and include variables {{like_this}}",
+        );
```
I chained the should statements here instead, so it looks more connected.

JesseSDevaney commented on e2e/test/scenarios/dashboard/text-dashcards.cy.spec.js:
```diff
@@ -0,0 +1,202 @@
+
+      // should auto-preview typed text
+      cy.get("main").findByText("You're editing this dashboard.").click(); // un-focus heading
+      getDashboardCard(1).within(() => {
+        cy.get("div").contains("Text text text");
+      });
```
findByText was not working here because react markdown styles the text with <em> and <strong> tags

JesseSDevaney committed:
refactor e2e tests

JesseSDevaney committed:
rename styling variables

JesseSDevaney committed:
null coalesce minWidth

JesseSDevaney committed:
remove unnecessary anonymous function wrapping

JesseSDevaney committed:
refactor to useToggle

JesseSDevaney committed:
improve no content check

JesseSDevaney committed:
use object creation shorthand

JesseSDevaney committed:
simplify expression

JesseSDevaney committed:
refactor empty content styling

JesseSDevaney committed:
move text-edit-container styling into heading and text components

JesseSDevaney committed:
refactor resize handle styling and abstract condition into function

JesseSDevaney committed:
refactor to use CSS function for string styling

JesseSDevaney committed:
improve Heading unit tests

JesseSDevaney committed:
improve Text unit tests

JesseSDevaney committed:
update Heading and Text unit tests

JesseSDevaney commented:
Hi @alxnddr @iethree!

:heart: Thank you for all of the feedback! I appreciate your help improving this feature. I just pushed all of the changes up based on your suggestions.


---
**Actionable Feedback for My Next PRs**
_Please add anything you think is important that I may have missed_

- [ ] Split into smaller PRs
  - easier to engineer, write tests for, and review
- [ ] Re-use existing code
  - `useToggle` hook to manage the focus state
  - existing menus
- [ ] Be more specific with variable naming
  - e.g. 
    - `showVisualizationOptions` -> `showDashcardVisualizationSettings`
    - `shouldAutoPreviewInEditing` - > `shouldForceHiddenBackground`
- [ ] Query elements by role and label instead of icons and text
  - e.q.
    - `cy.icon("pencil")` -> `cy.findByLabelText("Edit dashboard")`
    - `cy.icon("string)` -> `cy.findByLabelText("Add a heading or text box")`
    - `cy.get(".Modal")` -> `cy.findByRole("dialog")`
- [ ] Prefer longer flows over shorter flows in E2E tests (expensive to run in terms of time and resources) and leave smaller flows for unit tests
- [ ] Prefer styling components themselves instead of global CSS
  - Prefer REM units with steps of 0.25

JesseSDevaney committed:
Merge branch 'master' into new-dashboard-primitive-headings-prototype

JesseSDevaney committed:
update for new Icons

JesseSDevaney committed:
fix E2E tests

iethree commented on frontend/src/metabase/visualizations/visualizations/Heading/Heading.unit.spec.tsx:
```diff
@@ -0,0 +1,123 @@
+        };
+        setup(options);
+
+        screen.getByTestId("editing-dashboard-heading-preview").click();
```
It's not clear to me whether the chained `.click()` here uses fireEvent or userEvent 🤔  

Where possible, we prefer to use the `userEvent` library for testing user interactions in user tests. It puts a slightly higher level of abstraction that, in many cases, more accurately simulates user interactions than firing raw events. 

https://testing-library.com/docs/user-event/intro

JesseSDevaney replied:
Thank you for this!

iethree reviewed (approved):
Looks good to me! 🥳

alxnddr reviewed (approved):
Great job 🏅

JesseSDevaney committed:
refactor E2E test assertions

JesseSDevaney committed:
refactor click events to use user-event library

kdoh reviewed (dismissed):
Looking great overall. 🥳  Found one tiny polish thing I'd love if we could fix around the Tooltip for the new `Add a heading or text` menu in the edit state.

The tooltip seems to be attached to an element inside the action element bounds just for the `EntityMenu` here. In all the other actions it's attached correctly to the overall element rather than something inside. 
<img width="257" alt="Screen Shot 2023-06-08 at 12 13 57 PM" src="https://github.com/metabase/metabase/assets/5248953/77e9d338-c2b7-4f93-8b6b-d2ac463a91ef">
<img width="245" alt="Screen Shot 2023-06-08 at 12 13 51 PM" src="https://github.com/metabase/metabase/assets/5248953/1fd21962-bf56-410a-afeb-b36bcde7ceed">

JesseSDevaney committed:
fix tooltip positioning

JesseSDevaney commented:
Thanks @kdoh! 

Fixed:
![image](https://github.com/metabase/metabase/assets/22608765/647ddec4-d580-4dfa-b11b-6480fafa7258)
